src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractExtensionMethodImportCompletionProvider.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions.ContextQuery;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal abstract class AbstractExtensionMethodImportCompletionProvider : AbstractImportCompletionProvider
+    {
+        protected abstract string GenericSuffix { get; }
+
+        protected override bool ShouldProvideCompletion(Document document, SyntaxContext syntaxContext)
+            => syntaxContext.IsRightOfNameSeparator && IsAddingImportsSupported(document);
+
+        protected async override Task AddCompletionItemsAsync(
+            CompletionContext completionContext,
+            SyntaxContext syntaxContext,
+            HashSet<string> namespaceInScope,
+            bool isExpandedCompletion,
+            CancellationToken cancellationToken)
+        {
+            var syntaxFacts = completionContext.Document.Project.LanguageServices.GetRequiredService<ISyntaxFactsService>();
```

Comment:
why don't we have GetRequiredLanguageService as an extension on Document?

Reply:
Added